### DISCUSSION TITLE
How to get CallExpression arguments with esquery selector?

### DISCUSSION THREAD

**Discussion by MaksimShymko** [eslint/eslint#15553] [Q&A] [answered]
I'm trying to use `no-restricted-syntax` ESLint rule. The goal is to select MemberExpression except CallExpression callee.
Example:
```js
const obj = {
    val1: 1,
    val2: 2,
    funcParent: {
      func: () => {},
    },
};
const b = obj.funcParent.func(obj.val1, obj.val2); // select obj.val1 and obj.val2, but not obj.funcParent.func
```
My attempts:
- `CallExpression > MemberExpression` - selects `obj.funcParent.func`, `obj.val1`, `obj.val2`
- `CallExpression.arguments > MemberExpression` - selects nothing
- `CallExpression > MemberExpression:not(:first-child)` - selects `obj.funcParent.func`, `obj.val2`

How to select only CallExpression arguments (in our case `obj.val1` and `obj.val2`)?

**Comment by mdjermanovic** [comment, marked as answer]
> How to select only CallExpression arguments (in our case `obj.val1` and `obj.val2`)?

`CallExpression > MemberExpression.arguments` selects MemberExpression nodes that are CallExpression arguments.

[ESLint Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCBcIkNhbGxFeHByZXNzaW9uID4gTWVtYmVyRXhwcmVzc2lvbi5hcmd1bWVudHNcIl0gKi9cblxuY29uc3QgYiA9IG9iai5mdW5jUGFyZW50LmZ1bmMob2JqLnZhbDEsIG9iai52YWwyKTsgLy8gc2VsZWN0IG9iai52YWwxIGFuZCBvYmoudmFsMiwgYnV0IG5vdCBvYmouZnVuY1BhcmVudC5mdW5jIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Reply by MaksimShymko** [reply]
Could you explain why you get `arguments` on MemberExpression?

**Reply by MaksimShymko** [reply]
@mdjermanovic is it also possible to avoid that case? [Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCBcIkNhbGxFeHByZXNzaW9uID4gTWVtYmVyRXhwcmVzc2lvbi5hcmd1bWVudHNcIl0gKi9cblxuY29uc3QgYiA9IG9iai5mdW5jUGFyZW50LmZ1bmMoe2E6IG9iai52YWwxfSwge2I6IG9iai52YWwyfSk7IC8vIHNlbGVjdCBvYmoudmFsMSBhbmQgb2JqLnZhbDIsIGJ1dCBub3Qgb2JqLmZ1bmNQYXJlbnQuZnVuYyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Reply by mdjermanovic** [reply]
> Could you explain why you get `arguments` on MemberExpression?

`.foo` matches all nodes that are `foo` property of their parent node, or are in the `foo` array if parent's `foo` is an array. It's a path syntax that originates from [esquery](https://github.com/estools/esquery).

`Bar.foo` is a compound selector that matches all nodes that are matched by both `Bar` (node type) and `.foo`.

`Baz > Bar.foo` matches all nodes that are matched by `Bar.foo` and have `Baz` as the parent node.

> @mdjermanovic is it also possible to avoid that case? [Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCBcIkNhbGxFeHByZXNzaW9uID4gTWVtYmVyRXhwcmVzc2lvbi5hcmd1bWVudHNcIl0gKi9cblxuY29uc3QgYiA9IG9iai5mdW5jUGFyZW50LmZ1bmMoe2E6IG9iai52YWwxfSwge2I6IG9iai52YWwyfSk7IC8vIHNlbGVjdCBvYmoudmFsMSBhbmQgb2JqLnZhbDIsIGJ1dCBub3Qgb2JqLmZ1bmNQYXJlbnQuZnVuYyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

`CallExpression > MemberExpression.arguments` matches only direct children of `CallExpression`, because of `>`. If you would like to match any MemberExpression descendants of CallExpression arguments, then it would be `CallExpression > .arguments MemberExpression`
